File: apps/api/app/routers/model_catalog.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.core.deps import get_current_user, get_db_session
from app.core.errors import ApiError
from app.models import ModelCatalog, User
from app.schemas.model_catalog import ModelCatalogDetailOut, ModelCatalogOut

router = APIRouter(prefix="/api/v1/models/catalog", tags=["model-catalog"])

PROVIDER_DISPLAY_NAMES = {
    "qwen": "千问 · 阿里云",
    "alibaba": "千问 · 阿里云",
    "deepseek": "DeepSeek",
}

MODEL_DETAIL_OVERRIDES: dict[str, dict[str, object]] = {
    "qwen3.5-flash": {
        "supports_structured_output": True,
        "supports_cache": True,
    },
    "qwen3.5-plus": {
        "supports_structured_output": True,
        "supports_cache": True,
    },
    "qwen3-max": {
        "supports_structured_output": True,
        "supports_cache": True,
    },
    "deepseek-v3.2": {
        "supports_structured_output": True,
        "supports_cache": True,
    },
    "deepseek-r1": {
        "supports_structured_output": False,
        "supports_cache": False,
    },
    "qwen3-omni-flash-realtime": {
        "supports_structured_output": False,
        "supports_cache": False,
    },
    "paraformer-v2": {
        "price_unit": "audio",
        "price_note": "免费额度",
    },
    "sensevoice-v1": {
        "price_unit": "audio",
        "price_note": "免费额度",
    },
    "cosyvoice-v1": {
        "price_unit": "characters",
        "price_note": "按字符计费",
    },
    "sambert-v1": {
        "price_unit": "characters",
        "price_note": "按字符计费",
    },
}


def _provider_display_name(provider: str) -> str:
    key = (provider or "").lower()
    for prefix, label in PROVIDER_DISPLAY_NAMES.items():
        if key.startswith(prefix) or prefix in key:
            return label
    return provider


def _derive_modalities(item: ModelCatalog) -> tuple[list[str], list[str]]:
    capabilities = set(item.capabilities or [])

    if item.category == "llm":
        input_modalities = ["text"]
        if "vision" in capabilities:
            input_modalities.append("image")
        if "video" in capabilities:
            input_modalities.append("video")
        if "audio_input" in capabilities:
            input_modalities.append("audio")
        output_modalities = ["text"]
        if "audio_output" in capabilities:
            output_modalities.append("audio")
        return input_modalities, output_modalities

    if item.category == "asr":
        return ["audio"], ["text"]

    if item.category == "tts":
        return ["text"], ["audio"]

    if item.category == "vision":
        input_modalities = ["image"]
        if "video" in capabilities:
            input_modalities.append("video")
        return input_modalities, ["text"]

    return ["text"], ["text"]


def _build_catalog_detail(item: ModelCatalog) -> ModelCatalogDetailOut:
    capabilities = set(item.capabilities or [])
    overrides = MODEL_DETAIL_OVERRIDES.get(item.model_id, {})
    input_modalities, output_modalities = _derive_modalities(item)

    supports_function_calling = "function_calling" in capabilities
    supports_web_search = "web_search" in capabilities
    supports_structured_output = bool(
        overrides.get(
            "supports_structured_output",
            item.category == "llm" and "reasoning_chain" not in capabilities,
        )
    )
    supports_cache = bool(
        overrides.get(
            "supports_cache",
            item.category == "llm" and "realtime" not in capabilities,
        )
    )

    return ModelCatalogDetailOut(
        **ModelCatalogOut.model_validate(item, from_attributes=True).model_dump(),
        provider_display=_provider_display_name(item.provider),
        input_modalities=input_modalities,
        output_modalities=output_modalities,
        supports_function_calling=supports_function_calling,
        supports_web_search=supports_web_search,
        supports_structured_output=supports_structured_output,
        supports_cache=supports_cache,
        batch_input_price=overrides.get("batch_input_price"),
        batch_output_price=overrides.get("batch_output_price"),
        cache_read_price=overrides.get("cache_read_price"),
        cache_write_price=overrides.get("cache_write_price"),
        price_unit=str(overrides.get("price_unit", "tokens")),
        price_note=overrides.get("price_note"),
    )


@router.get("", response_model=list[ModelCatalogOut])
def list_catalog(
    category: str | None = Query(default=None),
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user),
) -> list[ModelCatalogOut]:
    _ = current_user
    query = db.query(ModelCatalog).filter(ModelCatalog.is_active.is_(True))
    if category:
        query = query.filter(ModelCatalog.category == category.lower())
    items = query.order_by(ModelCatalog.sort_order).all()
    return [ModelCatalogOut.model_validate(item, from_attributes=True) for item in items]


@router.get("/{model_id}", response_model=ModelCatalogDetailOut)
def get_catalog_item(
    model_id: str,
    db: Session = Depends(get_db_session),
    current_user: User = Depends(get_current_user),
) -> ModelCatalogDetailOut:
    _ = current_user
    item = (
        db.query(ModelCatalog)
        .filter(
            or_(ModelCatalog.model_id == model_id, ModelCatalog.id == model_id),
            ModelCatalog.is_active.is_(True),
        )
        .first()
    )
    if not item:
        raise ApiError("not_found", "Model not found in catalog", status_code=404)
    return _build_catalog_detail(item)

File: apps/web/lib/model-catalog.ts
const MODEL_CATEGORIES = ["llm", "asr", "tts", "vision"] as const;

export type ModelCategory = (typeof MODEL_CATEGORIES)[number];

export interface CatalogModelSummary {
  id: string;
  model_id: string;
  display_name: string;
  provider: string;
  category: ModelCategory;
  description: string;
  capabilities: string[];
  input_price: number;
  output_price: number;
  context_window: number;
  max_output: number;
}

export interface CatalogModelDetail extends CatalogModelSummary {
  provider_display: string;
  input_modalities: string[];
  output_modalities: string[];
  supports_function_calling: boolean;
  supports_web_search: boolean;
  supports_structured_output: boolean;
  supports_cache: boolean;
  batch_input_price: number | null;
  batch_output_price: number | null;
  cache_read_price: number | null;
  cache_write_price: number | null;
  price_unit: string;
  price_note: string | null;
}

type LooseCatalogModel = Partial<CatalogModelDetail> & Record<string, unknown>;

function asString(value: unknown, fallback = ""): string {
  return typeof value === "string" ? value : fallback;
}

function asNumber(value: unknown, fallback = 0): number {
  return typeof value === "number" && Number.isFinite(value) ? value : fallback;
}

function asOptionalNumber(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function asStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value.filter((item): item is string => typeof item === "string" && item.length > 0);
}

function inferCategory(raw: LooseCatalogModel, capabilities: string[]): ModelCategory {
  const category = asString(raw.category);
  if (MODEL_CATEGORIES.includes(category as ModelCategory)) {
    return category as ModelCategory;
  }

  const modelId = asString(raw.model_id).toLowerCase();
  const loweredCaps = capabilities.map((cap) => cap.toLowerCase());

  if (
    loweredCaps.includes("chat")
    || modelId.includes("qwen")
    || modelId.includes("deepseek")
  ) {
    return "llm";
  }
  if (
    modelId.includes("paraformer")
    || modelId.includes("sensevoice")
    || loweredCaps.includes("asr")
  ) {
    return "asr";
  }
  if (
    modelId.includes("cosyvoice")
    || modelId.includes("sambert")
    || loweredCaps.includes("tts")
  ) {
    return "tts";
  }
  if (
    modelId.includes("vision")
    || modelId.includes("ocr")
    || modelId.includes("internvl")
    || loweredCaps.includes("vision")
    || loweredCaps.includes("video")
  ) {
    return "vision";
  }

  return "llm";
}

function deriveModalities(category: ModelCategory, capabilities: string[]) {
  const loweredCaps = capabilities.map((cap) => cap.toLowerCase());

  if (category === "llm") {
    const input_modalities = ["text"];
    if (loweredCaps.includes("vision")) {
      input_modalities.push("image");
    }
    if (loweredCaps.includes("audio_input")) {
      input_modalities.push("audio");
    }
    const output_modalities = ["text"];
    if (loweredCaps.includes("audio_output")) {
      output_modalities.push("audio");
    }
    return { input_modalities, output_modalities };
  }

  if (category === "asr") {
    return { input_modalities: ["audio"], output_modalities: ["text"] };
  }

  if (category === "tts") {
    return { input_modalities: ["text"], output_modalities: ["audio"] };
  }

  const input_modalities = ["image"];
  if (loweredCaps.includes("video")) {
    input_modalities.push("video");
  }
  return { input_modalities, output_modalities: ["text"] };
}

export function normalizeCatalogModelSummary(raw: LooseCatalogModel): CatalogModelSummary {
  const capabilities = asStringArray(raw.capabilities);
  const model_id = asString(raw.model_id) || asString(raw.id) || "unknown-model";
  const category = inferCategory(raw, capabilities);

  return {
    id: asString(raw.id) || model_id,
    model_id,
    display_name: asString(raw.display_name) || model_id,
    provider: asString(raw.provider) || "mingrun",
    category,
    description: asString(raw.description),
    capabilities,
    input_price: asNumber(raw.input_price),
    output_price: asNumber(raw.output_price),
    context_window: asNumber(raw.context_window),
    max_output: asNumber(raw.max_output),
  };
}

export function normalizeCatalogModelDetail(raw: LooseCatalogModel): CatalogModelDetail {
  const summary = normalizeCatalogModelSummary(raw);
  const capabilities = summary.capabilities;
  const category = summary.category;
  const modalities = deriveModalities(category, capabilities);

  return {
    ...summary,
    provider_display: asString(raw.provider_display) || summary.provider,
    input_modalities: asStringArray(raw.input_modalities).length > 0
      ? asStringArray(raw.input_modalities)
      : modalities.input_modalities,
    output_modalities: asStringArray(raw.output_modalities).length > 0
      ? asStringArray(raw.output_modalities)
      : modalities.output_modalities,
    supports_function_calling: Boolean(raw.supports_function_calling ?? capabilities.includes("function_calling")),
    supports_web_search: Boolean(raw.supports_web_search ?? capabilities.includes("web_search")),
    supports_structured_output: Boolean(raw.supports_structured_output),
    supports_cache: Boolean(raw.supports_cache),
    batch_input_price: asOptionalNumber(raw.batch_input_price),
    batch_output_price: asOptionalNumber(raw.batch_output_price),
    cache_read_price: asOptionalNumber(raw.cache_read_price),
    cache_write_price: asOptionalNumber(raw.cache_write_price),
    price_unit: asString(raw.price_unit) || (category === "asr" ? "audio" : category === "tts" ? "characters" : "tokens"),
    price_note: asString(raw.price_note) || null,
  };
}
